web-api/openapi/schemas/user.yml

```diff
@@ -2,3 +2,23 @@ UserId:
   description: Id of a user.
   $ref: './id.yml#/Id'
   example: "user_id"
+
+InputUser:
+  description: |
+    Information about a user provided as input for an search
+
+    This can be anything from no user information, to a user id, to
+    part of the user data inlined.
+
+    Be aware that currently using the `id` in combination with any
+    other field is not allowed.
+  type: object
+  properties:
+    id:
+      $ref: '#/UserId'
+    history:
+      $ref: './document.yml#/History'
+  # Disabled as our linter can't handle that correctly
```

Comment:
I don't think `oneOf` with that shape is valid at all w.r.t. the specification.

```yaml
    InputUserById:
      type: object
      required:
        - id
      properties:
        id:
          type: string
    InputUserByHistory:
      type: object
      required:
        - history
      properties:
        history:
          type: string
    InputUser:
      oneOf:
        - $ref: '#/components/schemas/InputUserById'
        - $ref: '#/components/schemas/InputUserByHistory'
```

This should do what we want.

Reply:
We should also move `exclude_seen` under `user` since it does not make sense without it.

```yaml
InputUserById:
  type: object
    required:
      - id
    properties:
      id:
        type: string
InputUserByHistory:
  type: object
    required:
      - history
    properties:
      history:
        type: string
InputUserBy:
  oneOf:
    - $ref: '#/InputUserById'
    - $ref: '#/InputUserByHistory'

InputUser:
  type: object
  allOf:
    - $ref: '#/InputUserBy'
    - type: object
      properties:
        exclude_seen:
           type: bool
```

Reply:
> I don't think oneOf with that shape is valid at all w.r.t. the specification.

No it is valid, it's just the tooling which doesn't work correctly.

Json schema is not based on describing data but describing _validation rules_, and `oneOf` is a keyword which "tests" the data against the object it's contained in with each of it's variants inlined into it once and makes sure it only matches once.

(The  relevant part is here https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-10.2, the first line "These keywords apply subschemas to the same location in the instance as the parent schema is being applied.")

I originally had a version similar to what you proposed but decided against it as it's semantically not quite right.

We don't have multiple different entities you can pass in as the user fields but just one (partial) inlined "user" entity. Of which we can include various fields, like the id, history, center of interest etc. Currently we are limited to passing in id and history and currently we allow only either of them. But that's  more an implementation detail then anything else. I just for now don't want to handle questions like "do you store the inline passed in history", "do you recreate CoIs from the inline history or do you update the users cois", "do you store the computed cois", etc.  So it's more an limitation due to  implementation details then a fundamental aspect of the endpoint.  

For the same reason I didn't include `exclude_seen` in the user entity/object, it's not an property of the user but the endpoint where we can pass in a user "inline".

Reply:
You know what I will change it.

I originally had the idea that we could long term just provide an arbitrary subset of the user object and then act appropriately, i.e. if id fetch data from db and store it back, if inline history update/create cois from it, if inline cois use them as a base etc.

But just doing the oneOf like you wrote is simpler and more in line with how it's currently implemented.

Reply:
Through I will still keep the `exclud_seen` out of the user struct.

This: 

```
InputUser:
  type: object
  allOf:
    - oneOf:
      - $ref: '#/UserById'
      - $ref: '#/UserByHistory'
    - type: object
      properties:
        exclude_seen:
          type: boolean
```

is nearly guaranteed to trip over some tools :see_no_evil:

Reply:
So I found out when I tried I put the `oneOf` in the wrong place and I was getting an error also from other openapi tools.

> We don't have multiple different entities you can pass in as the user fields but just one (partial) inlined "user" entity. Of which we can include various fields, like the id, history, center of interest etc.

I disagree with this. In the context of this request, it is totally valid to see the user as an entity with an id or an entity with history.
Most of the problem about the semantics seems to come from having the field called `user`.
We can call it `personalize`, and I think most of the problems will go away:
```yaml
personalize:
  type object
  properties:
    exclude_seen:
      type: boolean
    for: # other option for the name: with, using, user not a strong opinion on this
      type object
      oneOf:
        - $ref: '#/FromId'
        - $ref: '#/FromHistory'
```
What do you think about this?

Reply:
okay